File: src/main/java/ru/TlkFile.java
package ru;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.w3c.dom.Document;
import org.xml.sax.InputSource;
import org.xml.sax.SAXException;

import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;
import javax.xml.stream.XMLOutputFactory;
import javax.xml.stream.XMLStreamException;
import javax.xml.stream.XMLStreamWriter;
import javax.xml.transform.OutputKeys;
import javax.xml.transform.Transformer;
import javax.xml.transform.TransformerException;
import javax.xml.transform.TransformerFactory;
import javax.xml.transform.dom.DOMSource;
import javax.xml.transform.stream.StreamResult;
import java.io.*;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

public class TlkFile  {

    private TlkHeader header;
    private List<TlkStringRef> stringRefs;
    private List<HuffmanNode> characterTree;
    private BitArray bits;

    /** <summary>
     *      Loads a TLK file into memory.
     * </summary>
     * <param name="fileName"></param>
     */
    public void loadTlkData(String fileName, boolean isPC) throws IOException {
        long start = System.nanoTime();
        /** **************** STEP ONE ****************
         *          -- load TLK file header --
         *        reading first 28 (4 * 7) bytes
         */

        /** using LittleEndian for PC architecture and BigEndian for Xbox360 */
        PositionInputStream r = baseStreamSeek( 0, fileName);

        header = new TlkHeader(r);
        /** read possibly correct ME2 TLK file, but from another platfrom */
        if (header.magic == 1416391424) {
            throw new RuntimeException();
        }
        /** read definitely NOT a ME2 TLK ile */
        if (header.magic != 7040084) {
            throw new RuntimeException();
        }
        //DebugTools.PrintHeader(Header);

        /** **************** STEP TWO ****************
         *  -- read and store Huffman Tree nodes --
         */
        /** jumping to the beginning of Huffmann Tree stored in TLK file */
        long pos = r.getPosition();
        r = baseStreamSeek( pos + (header.entry1Count + header.entry2Count)*8L, fileName);

        characterTree = new LinkedList<>();
        for (int i = 0; i < header.treeNodeCount; i++) {
            characterTree.add(new HuffmanNode(r));
        }

        /** **************** STEP THREE ****************
         *  -- read all of coded data into memory --
         */
        byte[] data = new byte[header.dataLen];
        r.read(data,0, data.length);
        /** and store it as raw bits for further processing */
        bits = new BitArray(data);
        /** rewind BinaryReader just after the Header
         * at the beginning of TLK Entries data
         * */
        r = baseStreamSeek(pos, fileName);

        /** **************** STEP FOUR ****************
         * -- decode (basing on Huffman Tree) raw bits data into actual strings --
         * and store them in a Dictionary<int, string> where:
         *   int: bit offset of the beginning of data (offset starting at 0 and counted for Bits array)
         *        so offset == 0 means the first bit in Bits array
         *   string: actual decoded String
         */
        Map<Integer, String> rawStrings = new HashMap<>();
        int offset = 0;
        Wrap offsetWrap = new Wrap(offset);
        while (offsetWrap.getValue() < bits.length()) {
            int key = offsetWrap.getValue();
            /** read the String and update 'offset' variable to store NEXT String offset */
            String s = GetString(offsetWrap);
            rawStrings.put(key, s);
        }

        /** **************** STEP FIVE ****************
         *         -- bind data to String IDs --
         * go through Entries in TLK file and read it's String ID and offset
         * then check if offset is a key in rawStrings and if it is, then bind data.
         * Sometimes there's no such key, in that case, our String ID is probably a substring
         * of another String present in rawStrings.
         */
        stringRefs = new LinkedList<>();
        for (int i = 0; i < header.entry1Count + header.entry2Count; i++) {
            TlkStringRef sRef = new TlkStringRef(r);
            sRef.position = i;
            if (sRef.bitOffset >= 0) {
                /** actually, it should store the fullString and subStringOffset,
                 * but as we don't have to use this compression feature,
                 * we will store only the part of String we need
                 * */
                /** int key = rawStrings.Keys.Last(c => c < sRef.BitOffset);
                 * String fullString = rawStrings[key];
                 * int subStringOffset = fullString.LastIndexOf(partString);
                 * sRef.StartOfString = subStringOffset;
                 * sRef.Data = fullString;
                 */
                sRef.data = rawStrings.containsKey(sRef.bitOffset) ?
                    rawStrings.get(sRef.bitOffset) :
                    GetString(new Wrap(sRef.bitOffset));
            }
            stringRefs.add(sRef);
        }
        r.close();
        System.out.println("loadTlkData: "+ (System.nanoTime() - start)/1_000_000);
    }

    /** <summary>
     *      Writes data stored in memory to an appriopriate text format.
     *  </summary>
     *  <param name="fileName"></param>
     *  <param name="ff"></param>
     */
    public void storeToFile(String fileName, FileFormat ff, FxmlController controller) throws XMLStreamException, IOException {

        long start = System.nanoTime();
        Files.deleteIfExists(Paths.get(fileName));
        /** for now, it's better not to sort, to preserve original order */
        // StringRefs.Sort(CompareTlkStringRef);

        if (ff.equals(FileFormat.XML)) {
            saveToXmlFile(fileName, controller);
            prettyXmlFile(fileName);
        } else {
            saveToTextFile(fileName, controller);
        }

        System.out.println("storeToFile: "+ (System.nanoTime() - start)/1_000_000);
    }

    /** <summary>
     *      Starts reading 'Bits' array at position 'bitOffset'. Read data is
     *      used on a Huffman Tree to decode read bits into real strings.
     *      'bitOffset' variable is updated with last read bit PLUS ONE (first unread bit).
     *  </summary>
     *  <param name="bitOffset"></param>
     *  <returns>
     *      decoded String or null if there's an error (last string's bit code is incomplete)
     *  </returns>
     *  <remarks>
     *      Global variables used:
     *      List(of HuffmanNodes) CharacterTree
     *      main.java.ru.BitArray Bits
     *  </remarks>
     */
    private String GetString(Wrap bitOffsetWrap) {
        HuffmanNode root = characterTree.get(0);
        HuffmanNode curNode = root;
        String curString = "";
        int i;
        for (i = bitOffsetWrap.getValue(); i < bits.length(); i++) {
            /** reading bits' sequence and decoding it to Strings while traversing Huffman Tree */
            int nextNodeID = bits.getRev(i) ? curNode.rightNodeId : curNode.leftNodeId;

            if (nextNodeID >= 0) {
                /** it's an internal node - keep looking for a leaf */
                curNode = characterTree.get(nextNodeID);
            } else {
                /** it's a leaf! */
                char c = 0;
                try {
                    c = BitConverter.toCharRev(BitConverter.GetBytes(0xffff - nextNodeID), 0);
                } catch (Exception e) {
                    e.printStackTrace();
                }
                if (c != '\0') {
                    /** it's not NULL */
                    curString += c;
                    curNode = root;
                } else {
                    /** it's a NULL terminating processed string, we're done */
                    i++;
                    bitOffsetWrap.setValue(i);
                    return curString;
                }
            }
        }
        i++;
        bitOffsetWrap.setValue(i);
        return null;
    }

    /** <summary>
     *       Writing data in an XML format.
     *  </summary>
     * <param name="fileName"></param>
     * */
    private void saveToXmlFile(String fileName, FxmlController controller) throws XMLStreamException, IOException {
        int totalCount = stringRefs.size();

        // Creating FileWriter object
        Writer fileWriter = new FileWriter(fileName);

        // Getting the XMLOutputFactory instance
        XMLOutputFactory xmlOutputFactory = XMLOutputFactory.newInstance();

        // Creating XMLStreamWriter object from xmlOutputFactory.
        XMLStreamWriter xr = xmlOutputFactory.createXMLStreamWriter(fileWriter);

        xr.writeStartDocument("utf-8", "1.0");
        xr.writeStartElement("tlkFile");
        xr.writeAttribute("TLKToolVersion", "1.0.4");
        xr.writeComment("Male entries section begin (ends at position " + (header.entry1Count - 1) + ")");

        for (int i = 0; i < stringRefs.size(); i++) {
            TlkStringRef s = stringRefs.get(i);
            if (s.position == header.entry1Count) {
                xr.writeComment("Male entries section end");
                xr.writeComment("Female entries section begin (ends at position " + (header.entry1Count + header.entry2Count - 1) + ")");
            }
            xr.writeStartElement("string");

            xr.writeStartElement("id");
            xr.writeCharacters(String.valueOf(s.stringId));
            xr.writeEndElement(); // </id>

            xr.writeStartElement("position");
            xr.writeCharacters(String.valueOf(s.position));
            xr.writeEndElement(); // </position>

            xr.writeStartElement("data");// </data>
            xr.writeCharacters(s.bitOffset < 0 ? "-1" : s.data);
            xr.writeEndElement(); // </data>

            xr.writeEndElement(); // </string>
        }
        xr.writeComment("Female entries section end");
        xr.writeEndElement(); // </tlkFile>
        xr.flush();
        xr.close();
    }

    private void prettyXmlFile(String fileName) throws IOException {
        String xmlString;
        try (BufferedReader reader = new BufferedReader(new FileReader(fileName))) {
            String line;
            StringBuilder stringBuilder = new StringBuilder();
            String ls = System.getProperty("line.separator");
            while ((line = reader.readLine()) != null) {
                stringBuilder.append(line);
                stringBuilder.append(ls);
            }
            xmlString = stringBuilder.toString();
        }

        try {
            InputSource src = new InputSource(new StringReader(xmlString));
            Document document = DocumentBuilderFactory.newInstance().newDocumentBuilder().parse(src);

            TransformerFactory transformerFactory = TransformerFactory.newInstance();
            transformerFactory.setAttribute("indent-number", 4);
            Transformer transformer = transformerFactory.newTransformer();
            transformer.setOutputProperty(OutputKeys.ENCODING, "UTF-8");
            transformer.setOutputProperty(OutputKeys.OMIT_XML_DECLARATION, "yes");
            transformer.setOutputProperty(OutputKeys.INDENT, "yes");

            PrintWriter writer = new PrintWriter(fileName);
            transformer.transform(new DOMSource(document), new StreamResult(writer));

        } catch (IOException | IllegalArgumentException |
                 ParserConfigurationException | TransformerException | SAXException e) {
            throw new RuntimeException("Error occurs when pretty-printing xml:\n" + xmlString, e);
        }
    }

    /** <summary>
     *      Writing data in a normal text format.
     * </summary>
     * <remarks>
     *      Currently not used by main application, but it works ok.
     * </remarks>
     * <param name="fileName"></param>
     */
    private void saveToTextFile(String fileName, FxmlController controller) {
        int totalCount = stringRefs.size();

        for (int i = 0; i < stringRefs.size(); i++) {
            TlkStringRef s = stringRefs.get(i);
            String line = s.stringId + ": " + s.data + "\r\n";

            try (FileWriter fw = new FileWriter(fileName, true)) {
                fw.write(line);
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    public static int readInt32(InputStream in) throws IOException {
        int ch1 = in.read();
        int ch2 = in.read();
        int ch3 = in.read();
        int ch4 = in.read();
        if ((ch1 | ch2 | ch3 | ch4) < 0) {
            throw new EOFException();
        }
        return (ch4 << 24) + (ch3 << 16) + (ch2 << 8) + ch1;
    }

    public static PositionInputStream baseStreamSeek(long pos, String fileName) throws IOException {
        PositionInputStream r = new PositionInputStream(Files.newInputStream(Paths.get(fileName)));
        r.skip(pos);
        return r;
    }

//        /** for sorting */
//        private static int CompareTlkStringRef(ru.TlkHeader.ru.TlkStringRef strRef1, ru.TlkHeader.ru.TlkStringRef strRef2) {
//            int result = strRef1.stringId.compareTo(strRef2.stringId);
//            return result;
//        }

    public static class TlkHeader {
        public int magic;
        public int ver;
        public int minVer;
        public int entry1Count;
        public int entry2Count;
        public int treeNodeCount;
        public int dataLen;

        public TlkHeader(InputStream r) throws IOException {
            this.magic = TlkFile.readInt32(r);
            this.ver = TlkFile.readInt32(r);
            this.minVer = TlkFile.readInt32(r);
            this.entry1Count = TlkFile.readInt32(r);
            this.entry2Count = TlkFile.readInt32(r);
            this.treeNodeCount = TlkFile.readInt32(r);
            this.dataLen = TlkFile.readInt32(r);
        }

    }

    public static class TlkStringRef {
        public int stringId;
        public int bitOffset;

        public String data;
        public int position;

        public TlkStringRef(InputStream r) throws IOException {
            this.stringId = TlkFile.readInt32(r);
            this.bitOffset = TlkFile.readInt32(r);
        }
    }

    public static class HuffmanNode {
        public int leftNodeId;
        public int rightNodeId;

        public HuffmanNode(InputStream r) throws IOException {
            this.leftNodeId = TlkFile.readInt32(r);
            this.rightNodeId = TlkFile.readInt32(r);
        }
    }

    @NoArgsConstructor @AllArgsConstructor
    @Getter @Setter
    public static class Wrap {
        int value;
    }

}

File: python_ver/tlk_file.py
import os

from bit_array import BitArray
import bit_convertor
from huffman_node import HuffmanNode
import tlk_header
from input_stream import InputStream
import tlk_string_ref
import offset_wrap

from xml.etree import ElementTree as ET


'''
Structure of TLK file of Mass Effect 2:
1. Header
    a) magic
    b) ver
    c) min_ver
    d) entry_1_count
    e) entry_2_count
    f) tree_nodes_count
    g) data_len
2. 
'''


class TlkFile:
    def __init__(self):
        self.header = None
        self.string_refs = []
        self.character_tree = []
        self.bits = BitArray()

    def get_string(self, bit_offset_wrap):
        root_node = self.character_tree[0]
        cur_node = root_node
        cur_string = ''

        i = bit_offset_wrap.value
        while i < self.bits.length:
            next_node_id = cur_node.right_node_id if self.bits.get_rev(i) else cur_node.left_node_id

            if next_node_id >= 0:
                cur_node = self.character_tree[next_node_id]
            else:
                try:
                    char = bit_convertor.to_char_rev(bit_convertor.get_bytes_by_value(0xffff - next_node_id), 0)
                except:
                    raise Exception
                if char != '\0':
                    cur_string += char
                    cur_node = root_node
                else:
                    i += 1
                    bit_offset_wrap.value = i
                    return cur_string
            i += 1
        i += 1
        bit_offset_wrap.value = i
        return None

    # Loads a TLK file into memory.
    def load_tlk_data(self, source_path):
        # ****************** STEP ONE ****************
        # -- load TLK file header --
        # reading first 28(4 * 7) bytes to build Header

        # using LittleEndian for PC architecture and BigEndian for Xbox360
        input_s = InputStream(source_path)
        self.header = tlk_header.TlkHeader(input_s)  # read 7 * 4 = 28 bytes

        # read possibly correct ME2 TLK file, but from another platfrom
        if self.header.magic == 1416391424:
            raise Exception('header.magic == 1416391424')
        # read definitely NOT a ME2 TLK ile
        if self.header.magic != 7040084:
            raise Exception('header.magic != 7040084')

        # ****************** STEP TWO ****************
        # -- read and store Huffman Tree nodes --

        # jumping to the beginning of Huffmann Tree stored in TLK file * /
        pos = input_s.pos  # position after reading of header
        input_s.pos = pos + (self.header.entry_1_count + self.header.entry_2_count) * 8  # TODO ???

        for _ in range(self.header.tree_nodes_count):
            h_node = HuffmanNode(input_s)  # read 8 bytes: 4 bytes to get left_node_id, 4 bytes to get right_node_id
            self.character_tree.append(h_node)

        # / ****************** STEP THREE ****************
        # -- read all of coded data into memory --
        data_length = self.header.data_len  # number of bytes of encoded sequence
        data = [None] * data_length
        input_s.read_to_array(data, 0, data_length)
        # and store it as raw bits for further processing
        self.bits = BitArray(a=data)

        # rewind BinaryReader just after the Header
        # at the beginning of TLK Entries data
        input_s.pos = pos

        # ****************** STEP FOUR ****************
        # -- decode (basing on Huffman Tree) raw bits data into actual strings --
        # and store them in a Dictionary <int,string> where:
        # int: bit offset of the beginning of data
        # (offset starting at 0 and counted for Bits array)
        # so offset == 0 means the first bit in Bits array
        # string: actual decoded String
        raw_str = {}
        offset = offset_wrap.OffsetWrap(0)  # just a wrapper of bits offset

        while offset.value < self.bits.length:
            key = offset.value
            s = self.get_string(offset)
            raw_str[key] = s

        # **************** STEP FIVE ****************
        # -- bind data to String IDs --
        # go through Entries in TLK file and read it's String ID and offset
        # then check if offset is a key in rawStrings and if it is, then bind data.
        # Sometimes there's no such key, in that case, our String ID is probably a substring
        # of another String present in rawStrings.
        for i in range(self.header.entry_1_count + self.header.entry_2_count):
            s_ref = tlk_string_ref.TlkStringRef(input_s)
            s_ref.position = i
            if s_ref.bit_offset >= 0:
                # actually, it should store the fullString and subStringOffset,
                # but as we don't have to use this compression feature,
                # we will store only the part of String we need

                # key = rawStrings.Keys.Last(c => c < sRef.BitOffset);
                # String fullString = raw_str[key];
                # int subStringOffset = fullString.LastIndexOf(partString);
                # s_ref.StartOfString = subStringOffset;
                # s_ref.Data = fullString;
                s_ref.data = raw_str[s_ref.bit_offset] \
                    if s_ref.bit_offset in raw_str.keys() \
                    else self.get_string(offset_wrap.OffsetWrap(s_ref.bit_offset))
            self.string_refs.append(s_ref)

    def store_to_file(self, dest_file: str, file_format: str):
        if os.path.isfile(dest_file):
            os.remove(dest_file)
        if file_format.lower() == 'to_xml':
            print(file_format + ' XML')
            self.save_to_xml_file(dest_file)
        else:
            print(file_format + ' TXT')
            self.save_to_text_file(dest_file)

    # Writing data in an XML format.
    def save_to_xml_file(self, abs_path):
        root = ET.Element('tlkFile')  # <tlkFile> tag
        root.set("TLKToolVersion", '1.0.4')  # <tlkFile> attributes
        comment = ET.Comment('Male entries section begin (ends at position {0})'.format(
            (str(self.header.entry_1_count - 1))))
        root.append(comment)

        for i in range(len(self.string_refs)):
            sr = self.string_refs[i]

            if sr.position == self.header.entry_1_count:
                comment1 = ET.Comment('Male entries section end')
                comment2 = ET.Comment('Female entries section begin (ends at position {0})'.format(
                    (str(self.header.entry_1_count + self.header.entry_2_count - 1))))
                root.append(comment1)
                root.append(comment2)

            s = ET.SubElement(root, 'string')  # <string> tag

            s1 = ET.SubElement(s, 'id')  # <id> tag
            s1.text = str(sr.string_id)

            s2 = ET.SubElement(s, 'position')  # <position> tag
            s2.text = str(sr.position)

            s3 = ET.SubElement(s, 'data')  # <data> tag
            s3.text = '-1' if sr.bit_offset < 0 else sr.data
        comment = ET.Comment("Female entries section end")
        root.append(comment)

        tree = ET.ElementTree(root)
        ET.indent(tree, space="\t", level=0)
        tree.write(abs_path, encoding="utf-8")

    def save_to_text_file(self, dest_file):
        total_count = len(self.string_refs)

        with open(dest_file, "w+") as f:
            for i in range(total_count):
                s = self.string_refs[i]
                line = str(s.string_id) + ': ' + str(s.data) + '\r\n'
                f.write(line)
